test(models): mocking database for unit test (#10)

BasixKOR opened the pull request:
This PR fixes #10.

mu-hun committed:
feat(post-processor): fetching school register data

TODO: refactor to polymorphic with a decorator pattern

BasixKOR committed:
chore(package): add cuid to emulate prisma id

BasixKOR committed:
test(model): use model mock in test

This mock currently assumes you to not make duplicate items with same
unique field. (such as name, etc)

BasixKOR committed:
test(models): make model to verify unique fields

mu-hun committed:
feat(fetch, hack?): semesters GPA parser

create 'src/@types/dreamy.ts' for dreamy.jejunu.ac.kr  API typing

BasixKOR commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,48 @@
+    ...input,
+    averagePoint: input.averagePoint || 4.2,
+    name: input.name || undefined,
+    id: input.id ? input.id.toString() : cuid(),
```
이 라인은 `UserCreateInput`과 `User` 타입이 선택 필드를 정의하는 방법이 조금 달라 해당 부분을 맞춰주는 workaround 코드입니다. 더 좋은 방법이 있으시다면 제안 부탁드려요 😄

mu-hun replied:
cuid를 서버 파트에 부담하고, `UserCreateInput`에 ID를 빼셔도 괜찮을 것 같습니다.

BasixKOR replied:
UserCreateInput은 Prisma 제공 타입이라 id를 빼는 건 어려울 것 같고, Prisma에서 id가 제공될 때 어떻게 작동하는지 잠깐 살펴보고 비슷하게 적용해볼게요

BasixKOR commented on src/index.test.ts:
```diff
@@ -12,6 +12,8 @@ interface TypeRes<T> extends request.Response {
 
 const invaild = 'INVAILD'
 
+jest.mock('./models/index')
```
위 코드가 있음에도 ESLint에서 __mocks__ 내 파일이 import 되지 않은 것으로 판단해 lint 중 오류를 냅니다. 향후 우회할 방법이 필요할 것으로 보입니다.

p.s. 커밋할 때에는 `--no-verify` 를 사용해 임의로 우회했습니다 😓

mu-hun replied:
임시로 문제가 나는 라인이나 파일 전체를 ignore 하면 괜찮지 않을까요?

mu-hun replied:
다시 보니 경로를`./models` 만 넣어도 될 것 같습니다.

mu-hun replied:
tsconfig의 include 속성에 해당 파일의 경로를 집어넣으니 해결되었습니다.

BasixKOR reviewed (commented):
일부 언급이 필요할 것 같은 부분에 대해 코멘트를 남깁니다.

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+    createdAt: '2019-10-29T14:53:17.524Z',
+    mailid: 'muhun',
+    name: '무훈',
+    averagePoint: 4.2,
```
제주대학교는 `3.5` 학점제를 사용하고 있습니다.

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+    id: 'ck2bz0s449a4v0919xqvuzht9',
+    createdAt: '2019-10-29T14:53:17.524Z',
+    mailid: 'muhun',
+    name: '무훈',
```
본래 파싱하는 데이터에선 성씨가 포함되어 있습니다.

BasixKOR replied:
수정하겠습니다.

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+  const user: User = {
+    createdAt: new Date().toISOString(),
+    ...input,
+    averagePoint: input.averagePoint || 4.2,
```
`input.averagePoint || 3.5` 입니다.

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+}
+
+export const isUserExist = async (mailid: string) =>
```
`user.some` 메서드는 `Promise` 를 반환하지 않습니다. `async` 를 빼야 됩니다.

BasixKOR replied:
Promise.resolve로 감싸야겠네요 😂

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+  },
+]
+
+export const createUser = async (input: UserCreateInput): Promise<User> => {
```
`await` 를 쓰는 구문이 없지만 [`src/checkVaild.ts#L34`](https://github.com/reflation/backend/blob/master/src/checkVaild.ts#L34)와의 호완성을 위해 놓아두신 건가요..?

BasixKOR replied:
Mocking하는 createUser가 Promise를 반환하므로 똑같이 작동하도록 구성했습니다.

mu-hun replied:
https://github.com/reflation/backend/pull/13#discussion_r341804809 답글과 동일하게 Promise.resolve 를 쓰겠습니다.

mu-hun reviewed (changes requested):
수정 요청 사안에 납득하신다면 제가 커밋을 해도 괜찮을까요?

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+    mailid: 'muhun',
+    name: '무훈',
+    averagePoint: 4.2,
+  },
```
다음 PR 작업 이전에 `semesters` 스키마를 User Type 에 적용해두면 좋겠네요.

mu-hun commented:
~~PS: `index.test.ts` [업데이트를 제안드립니다](https://gist.github.com/x86chi/f22e8be0eb351540a20506552a7c004b)~~ 5dfe465f734882eaadb64f9f35e9f3c723b041da 에서 반영함

수신 받은 body의 타입을 TypeUser과 비교하여([45줄 참고](https://gist.github.com/x86chi/f22e8be0eb351540a20506552a7c004b#file-index-test-ts-L45)) 체크할 좋은 방법이 있을까요?

mu-hun commented on src/models/__mocks__/index.ts:
```diff
@@ -0,0 +1,46 @@
+}: TypeUser): Promise<User> => {
+  const foundIndex = users.findIndex(user => user.mailid === mailid)
+  users[foundIndex].averagePoint = data || 4.2
+  return users[foundIndex]
```
```ts
  averagePoint,
}: TypeUser): Promise<User> => {
  const foundIndex = users.findIndex(user => user.mailid === mailid)
  users[foundIndex].averagePoint = averagePoint || 3.5
  return users[foundIndex]
}
```

`data`는 `TypeUser`에 없는 속성입니다.

mu-hun commented:
현재 피드백을 준 부분에 대한 커밋을 작업하고 있습니다.

mu-hun committed:
test(mock): apply change request from #13

side-effect: add mock ts file into include array of tsconfig

mu-hun committed:
merge(branch): 'school-register-to-prisma-client' into test/models-mock

Conflicts:
	src/routes.ts
	tsconfig.json

mu-hun committed:
test(mock): use promise in searchUser

mu-hun committed:
fix(mock): must use assign. not foreach

mu-hun committed:
refact: everything works fine

`@types/models.ts`: add `mailid` prop in `TypeUser`
`checkVailld.ts`, `routes.ts` :  when error occurred, return error props instead of throw syntax. fix #14
`index.test.ts` : missed ‘dotenv’ module, ‘POST /fetch’ must use form type.
`models/__mocks__/index.ts`:
1. Before unit test, mock data must be empty. resolve #14
2. use reject callback in appendUserData
`models/index.ts`, `utils/fetch.ts`: huge fix schema

close #13: now use JSON and form request, JSON response

Co-authored-by: Basix <basix@basix.tech>

mu-hun committed:
chore: change port number to 4000

this commit side-effect in 'test/models-mock' branch
